VERSIONING.md

```diff
@@ -0,0 +1,202 @@
+# Cosign Versioning Policy
+
+This is the versioning policy regarding the stability of Cosign, both the API and CLI.
+
+**tl;dr (CLI):** semver for specific behaviors, but with deprecations.
+
+**tl;dr (API):** no guarantees at all (for now).
+
+Principles:
+
+- Provide well-versioned, very reliable tools/libraries that people can write
+  scripts/code against.
+- Provide places for experimentation and rapid-development, not covered under
+  versioning guarantees.
+- Be explicit about what is and isn’t covered.
+- Better to promise a weaker guarantee and stick to it consistently than to be
+  vague about what is/isn’t covered and technically comply, or not comply at
+  all.
+
+## Cosign CLI
+
+The Cosign CLI’s version numbers are of the form `MAJOR.MINOR.PATCH`, but it is
+explicitly not covered by [semantic versioning][semver] guarantees. Instead, you
```

Comment:
As long as Cosign is going to cosplay as semver, it should be semver. If semver guarantees aren't going to be respected, then there should be an unambiguously (programmatically) different scheme.

Reply:
Not quite sure what you mean by "cosplay as semver." Where do you think the confusion will come from? The MAJOR.MINOR.PATCH scheme? IMO semver doesn't have a monopoly on that. Or just the fact that we have pretty-similar-but-not-quite-the-same guarantees?

I'd be open to other versioning schemes as long as they had he following features:

- support for big, all-compatibility-bets-are-off breaking version bumps
- version bumps corresponding to "typical" releases which can be breaking but only in accordance with a deprecation policy
- releases that will always be safe to upgrade to, and that we may backport fixes to

Seems like a 3-part scheme is pretty natural given those goals. We could try to do something with dates I guess? `v2.20230111` but then the date in a patch release if we're going to backport something is awkward. WDYT?